pypsa/optimization/constraints.py

```diff
@@ -415,40 +859,88 @@ def p_previous(idx):
             active.index, columns=fix_i
         )
         m.add_constraints(
-            lhs, ">=", rhs, name=f"{c}-fix-{attr}-ramp_limit_down", mask=mask
+            lhs, ">=", rhs, name=f"{c}-fix-non-comm-{attr}-ramp_limit_down", mask=mask
+        )
+
+    # ----------------------- Extendable but non committable Generators ----------------------------- #
+
+    ext_i = n.get_extendable_i(c)
+    com_i = n.get_committable_i(c)
+    ext_i = ext_i.difference(com_i).rename(ext_i.name)
+
+    assets = n.df(c).reindex(ext_i)
+
+    # ext up
+    if not ramp_limit_up[ext_i].isnull().all().all():
+        p_nom = m[f"{c}-p_nom"].loc[ext_i]
+        limit_pu = DataArray(ramp_limit_up.reindex(active.index, columns=ext_i))
+        lhs = p_actual(ext_i) - p_previous(ext_i) - limit_pu * p_nom
+        rhs = rhs_start.reindex(columns=ext_i)
+        mask = active.reindex(columns=ext_i) & ~ramp_limit_up.isnull().reindex(
+            active.index, columns=ext_i
+        )
+        m.add_constraints(
+            lhs, "<=", rhs, name=f"{c}-ext-non-comm-{attr}-ramp_limit_up", mask=mask
+        )
+
+    # ext down
+    if not ramp_limit_down[ext_i].isnull().all().all():
+        p_nom = m[f"{c}-p_nom"].loc[ext_i]
+        limit_pu = DataArray(ramp_limit_down.reindex(active.index, columns=ext_i))
+        lhs = p_actual(ext_i) - p_previous(ext_i) + limit_pu * p_nom
+        rhs = rhs_start.reindex(columns=ext_i)
+        mask = active.reindex(columns=ext_i) & ~ramp_limit_down.isnull().reindex(
+            active.index, columns=ext_i
+        )
+        m.add_constraints(
+            lhs, ">=", rhs, name=f"{c}-ext-non-comm-{attr}-ramp_limit_down", mask=mask
         )
 
-    # ----------------------------- Extendable Generators ----------------------------- #
+    # ----------------------- Extendable, committable but non modular Generators ----------------------------- #
+
+    ext_i = n.get_extendable_i(c)
+    com_i = n.get_committable_i(c)
+    not_mod_i = n.df(c).query(f"({nominal_attrs[c]}_mod==0)").index
+    ext_i = ext_i.intersection(com_i).intersection(not_mod_i).rename(ext_i.name)
```

Comment:
Given the frequent use of these block, we may create a n.get_extendable_and_commitable function maybe?

Reply:
Absolutely. If you think it could be useful, I can do it for sure :)